### ISSUE TITLE
Allow temporary virtual environments that are deleted after the build

### ISSUE THREAD

**Issue by vpratz** [real-yfprojects/sphinx-polyversion#17]
I'm not sure if this is in scope, and how many people will want to use this.

In a setting that is disk-space constrained, e.g. in runners like GitHub actions, having many (possibly large) virtual environments can become a problem. Running sequentially as proposed in #14 avoids creating all of them at the same time, but one also has to remove the virtual environments after a revision was built. Doing this manually is not hard (e.g. set up a function that purges the venv directory after each build), but one could also provide the functionality in `__aexit__`. This would for example look like this:

``` python
class TemporaryPip(Pip):
    async def __aexit__(self, exc_type, exc, tb):
        logger.info("Removing environment " + str(self.venv))
        shutil.rmtree(self.venv)
```

Automatically removing user-configured directories is a somewhat risky business, so we would probably want to safe-guard it in some way, e.g. by having a subdirectory `VENV_DIR_NAME=.sphinx-polyversion-venv-tmp` where we create the environents, and then using a check like this:

``` python
class TemporaryPip(DynamicPip):
    async def __aexit__(self, exc_type, exc, tb):
        if self.venv.parent.name == VENV_DIR_NAME:
            logger.info("Removing environment " + str(self.venv))
            shutil.rmtree(self.venv)
```

For the library, we would extend `Pip`, add a keyword argument like `temporary` and require a creator so that we can create and destroy temporary environments.

**Comment by real-yfprojects** [comment]
Did you encounter problem with disk space? Because `DefaultDriver.tmp_dir` creates a temporary directory which the virtual environment is placed in. After building a revision the temporary directory will be removed according to the [docs](https://docs.python.org/3.10/library/tempfile.html#tempfile.TemporaryDirectory). Therefore I would think that this addition isn't necessary.

**Comment by vpratz** [comment]
Thanks for the response and the explanation. The disk space problem arose because for `Pip` the virtual environments are created in a subfolder of my source directory, and not in the temporary environment.
I think the problem does not exist in `Poetry` because of [this line](https://github.com/real-yfprojects/sphinx-polyversion/blob/6b7a76320fb03b5e3a370bd3150f1856c83e0854/sphinx_polyversion/pyvenv.py#L267)

``` python
venv_path = self.path / ".venv"
```

where `self.path` is the temporary directory. In `VirtualPythonEnvironment`, `venv` is used without `path` [here](https://github.com/real-yfprojects/sphinx-polyversion/blob/6b7a76320fb03b5e3a370bd3150f1856c83e0854/sphinx_polyversion/pyvenv.py#L138), so this happens in a local directory and not a temporary one. Changing this behavior to match the one for `Sphinx` would solve the problem, as well as #16.
The would be that the environments cannot be cached anymore, even if disk space is not an issue. Also, having pre-existing environments without `creator` would not make sense in this context, where `venv` is always in a temporary directory. So maybe something configurable via a `temporary` flag would be useful to support both usages? Temporary would use `venv_path = self.path / venv`, whereas non-temporary would just use `venv_path = venv`. 
Do you agree with this assessment? What would be the best change to make here?

**Comment by vpratz** [comment]
Actually, I believe most of the problems I had setting this up can be traced to this behavior of `Pip`, which reading your comments is unintended. I think making making `creator` mandatory for `Pip` and creating the envs in the temporary directory (as you intended it) would be the most robust way.
The behavior to have local environments (if one wants to reuse them) is probably not so important, or, if needed, could be best met with a different class (e.g. `LocalEnvironment` or something). As pip and poetry cache the packages as well, caching the environment is probably not critical in most settings.

**Comment by real-yfprojects** [comment]
> Actually, I believe most of the problems I had setting this up can be traced to this behavior of `Pip`, which reading your comments is unintended. I think making making `creator` mandatory for `Pip` and creating the envs in the temporary directory (as you intended it) would be the most robust way.

Yeah, I guess we can do it similar to the `Poetry` environment. I didn't think this through when implementing the `Pip` class :).
